Refactor GC_ObjectModel into core OMR_GC_ObjectModel base class

ktbriggs opened the pull request:
[Continued from #296]

Split object model methods depending only on header slot offset and 
flags shift into core GC_ObjectModelBase base class to simplify the
process of constructing language-specfic object models. Removed 
spurious macro definitions. Client languages will extend this base 
class to produce their own GC_ObjectModel classes. 

Cleaned up example object model method comments. Updated Scavenger and
generational access barrier as required to adapt to object model 
changes.

Moved example/glue object flags byte to high order byte of object
header to allow some verification of base object model handling of 
header flags.

This is a first step toward setting example/glue/ObjectModel.[ch]pp up
as model/template to serve as starting point for development of 
language-specific object model classes. 

Issue: #296
Signed-off-by: Kim Briggs briggs@ca.ibm.com

ktbriggs commented:
This is ready for review but should not be merged into eclipse/omr until reviewed and accepted and also until dependent java commit has also been reviewed and accepted for merging. 

@amicic please review. ping me on slack if you would like to do a deskside review (recommended owing to the amount of refactoring done to GC_ObjectModel).

ktbriggs commented:
Personal cbuild for all platforms: http://vmfarm.ottawa.ibm.com/build_info.php?build_id=323307. One gold job _[Linux Hammer] 90 VM_Sanity.cmdLineTester_XcheckJNI.NoOptions.1_ failed (grinder failed 8/10); http://vmfarm.ottawa.ibm.com/build_info.php?build_id=323361. This test also fails when run on similar machine rhel6x64vm46 with SDKs from recent ibuilds 322260, 322313, 322451. And it is _not_ included as a gold job in production cbuilds or ibuilds; not sure why it is gold in personal cbuilds (jvm.29.personal).
```
Testing: memorycheck
Test start time: 2016/10/25 06:03:39 Eastern Standard Time
Running command: /bluebird/builds/bld_323307/sdk/xa6490/bin/java  -Xnocompressedrefs  -Xnocompressedrefs -cp "/bluebird/builds/bld_323307/jvmtest/VM/cmdline_options_testresources/cmdlinetestresources.jar" -Xcheck:jni -Xgcthreads1 -Xcheck:memory j9vm.test.jnichk.BufferOverrun
Time spent starting: 4 milliseconds
***[TEST INFO 2016/10/25 06:05:39] ProcessKiller detected a timeout after 120000 milliseconds!***
INFO: getUnixPID() has failed indicating this is not a UNIX System.'Debug on timeout' is currently only supported on Linux.
Output from test:
 [ERR] JVMJNCK001I 
Time spent executing: 125042 milliseconds
Test result: FAILED
Output from test:
 [ERR] scan_all_blocks describing block at 00007FCCE0B2E5B0 (header at 00007FCCE0B2E3B0):
 [ERR] First 1 bytes of bottom padding are damaged
```
Will discuss this with @DanHeidinga.

Further testing on rhel6x64vm46 (24 CPUs) running SPECjbb2005 in 8 stages, 7 30s warmup stages involving 3*n worker threads (0<n<8) and 4m peak stage involving 24 threads. For each gcpolicy.
```
j9build@rhel6x64vm46 /team/briggs $ runAllGC SPECjbb2005 --props SPECjbb-24CPU.props
Running SPECjbb2005 with optthruput in /team/briggs/jvmxa6490/lib/amd64/compressedrefs ...
/tmp/SPECjbb2005-rhel6x64vm46-0-jvmxa6490-optthruput-1025-073007.log
Score: 7368
Running SPECjbb2005 with optavgpause in /team/briggs/jvmxa6490/lib/amd64/compressedrefs ...
/tmp/SPECjbb2005-rhel6x64vm46-0-jvmxa6490-optavgpause-1025-073742.log
Score: 6582
Running SPECjbb2005 with gencon in /team/briggs/jvmxa6490/lib/amd64/compressedrefs ...
/tmp/SPECjbb2005-rhel6x64vm46-0-jvmxa6490-gencon-1025-074517.log
Score: 9850
Running SPECjbb2005 with metronome in /team/briggs/jvmxa6490/lib/amd64/compressedrefs ...
/tmp/SPECjbb2005-rhel6x64vm46-0-jvmxa6490-metronome-1025-075252.log
Score: 5878
Running SPECjbb2005 with balanced in /team/briggs/jvmxa6490/lib/amd64/compressedrefs ...
/tmp/SPECjbb2005-rhel6x64vm46-0-jvmxa6490-balanced-1025-080035.log
Score: 10425
```

charliegracie commented on gc/base/MarkedObjectPopulator.cpp:
```diff
@@ -72,9 +72,9 @@ MM_MarkedObjectPopulator::populateObjectHeapBufferedIteratorCache(omrobjectptr_t
+			state->data1 = ((uintptr_t)object) + OMR_MINIMUM_OBJECT_SIZE;
 #else /* OMR_GC_MINIMUM_OBJECT_SIZE */
-			state->data1 = ((uintptr_t)object) + sizeof(J9Object);
+			state->data1 = ((uintptr_t)object) + sizeof(fomrobject_t);
```
I believe this is incorrect.  I agree that sizeof(J9Object) is wrong and would result in a compile failures if OMR_GC_MINIMUM_OBJECT_SIZE is not defined.  This needs to move the pointer past the last object.  The macro sizeof(fomrobject_t) is just going to move it past either 32 or 64 bits which is likely not the minimum size of an object.

ktbriggs replied:
I have to look into this more closely. This change will have no effect on existing compiled code, since J9Object and fomrobject_t definitions are identical. I see your point (result should have object (8-byte) alignment, and I suspect that this is taken care of in the MM_ObjjectHeapBufferedIterator and its many Populator pals. Will verify this (or not).

ktbriggs replied:
This is correct as is. If OMR_GC_MINIMUM_OBJECT_SIZE is not defined the effective minimum object size is sizeof(fomrobject_t).

charliegracie commented on example/glue/ObjectModel.hpp:
```diff
@@ -97,425 +59,207 @@ class GC_ObjectModel
 	 */
 	MMINLINE uintptr_t
-	getPreservedFlags(MM_ForwardedHeader *forwardedHeader)
+	getConsumedSizeInBytesWithHeader(omrobjectptr_t objectPtr)
```
This feels like common code that belongs in the base class.

ktbriggs replied:
GC_ObjectModelBase has no way of determining object size. So all of the get*Size*With(out?)Header() methods have to live in subclasses. I was disappointed too.

ktbriggs replied:
I intend to address this in an upcoming commit -- will try using a template-based solution that will obviate need for virtual functions, eg, `class GC_ObjectModel<ObjectModelImplementation> {}` where `ObjectModelImplementation` implements `getObjectSizeWithoutHeader()`. Then GC_ObjectModel can implement the derivative `get(Consumed?)SizeIn(Bytes|Slots)With(out?)Header()` methods.

If that works out will refactor `MM_CollectorLanguageInterface::scavenger_*()` out into `GC_GenerationalCollectorInterface<GenerationalImplementation> {}` and so on for other components in CLI, GCLI, ELI.

We should touch base on this when you have a few minutes.

charliegracie commented on example/glue/ObjectModel.hpp:
```diff
@@ -97,425 +59,207 @@ class GC_ObjectModel
 	 */
 	MMINLINE uintptr_t
-	getPreservedAge(MM_ForwardedHeader *forwardedHeader)
+	getConsumedSizeInSlotsWithHeader(omrobjectptr_t objectPtr)
```
This feels like common code that belongs in the base class.

ktbriggs replied:
Same as previous comment -- base class can't do size.

charliegracie commented on example/glue/ObjectModel.hpp:
```diff
@@ -97,425 +59,207 @@ class GC_ObjectModel
 
 	/**
-	 * Extract the size from an unforwarded object.
+	 * See comment for getConsumedSizeInBytesWithHeader(), above. This method would be called
```
Is the function called for object movement in the scavenger?  If so the comment needs to be updated.

ktbriggs replied:
Not sure what needs to be updated in comment, which states that this function is called "just before the object is to be promoted into tenure space". So, in generational context, without specific reference to Scavenger.

charliegracie replied:
Sorry I meant for general flipping.  Objects may grow anytime they move just not when they are tenured.  They may also grow during compaction.

ktbriggs replied:
How about this? 
```
	 * See comment for getConsumedSizeInBytesWithHeader(), above. This method would be called
	 * just before the object is moved (eg, promoted into tenure space, or compacted) to 
	 * calculate the actual total size of the object in its current state plus additional bytes
	 * required to expand the object.
```

charliegracie commented on example/glue/ObjectModel.hpp:
```diff
@@ -538,30 +280,31 @@ class GC_ObjectModel
 	{
-		*objectCopySizeInBytes = forwardedHeader->getPreservedSlot() >> OMR_OBJECT_METADATA_SIZE_SHIFT;
+		*objectCopySizeInBytes = forwardedHeader->getPreservedSlot() & (((uintptr_t)1 << _objectMetadataSlotFlagsShift) - 1);
 		*objectReserveSizeInBytes = adjustSizeInBytes(*objectCopySizeInBytes);
```
This feels like common code that belongs in the base class.

ktbriggs replied:
See j9 GC_ObjectModel::calculateObjectDetailsForCopy(). There are no common factors here.

charliegracie commented on example/glue/ObjectModel.hpp:
```diff
@@ -570,95 +313,23 @@ class GC_ObjectModel
 	MMINLINE void
 	fixupForwardedObject(MM_ForwardedHeader *forwardedHeader, omrobjectptr_t destinationObjectPtr, uintptr_t objectAge)
```
This feels like common code that belongs in the base class.

ktbriggs replied:
The code that is common _is_ in the base class: 
```
GC_ObjectModelBase::fixupForwardedObject(forwardedHeader, destinationObjectPtr, objectAge);
```
is implementation for this function. In j9 GC_ObjectModel, base class method is called before implementing j9-specific fixup.

charliegracie commented on gc/base/AllocateInitialization.hpp:
```diff
@@ -175,8 +175,8 @@ class MM_AllocateInitialization : public MM_Base {
-			fomrobject_t *headerPtr = OMR_OBJECT_METADATA_SLOT_EA((omrobjectptr_t)heapBytes);
-			*headerPtr = (fomrobject_t)(_allocateDescription.getObjectFlags() << OMR_OBJECT_METADATA_FLAGS_SHIFT);
+			env->getExtensions()->objectModel.setFlags((omrobjectptr_t)heapBytes,
+					OMR_OBJECT_METADATA_FLAGS_MASK, _allocateDescription.getObjectFlags());
```
Why was this broken into 2 lines?

ktbriggs replied:
Anticipating giant code reformat. Not sure what we've landed on for line splits but this feels right.

ktbriggs commented:
@charliegracie suggest holding off on this and continuing on new pull/merge requests for omr/j9vm for rtc 119466 . Will prepare these shortly and omr will contain this commit.

ktbriggs commented:
@charliegracie This is ready for review but pull into eclipse/master must be put off pending template support in all build tooling or effective workaround for template use with existing tooling. 

This is a cookie-cutter design pattern that is robust and can be applied to existing language\configuration/collector/environment glue with ease. Each OMR component (eg, GC_ObjectModel) is represented as a base class template that receives a delegate class as template parameter and contains an instance of the delegate class. The methods in the delegate class can be inlined and obviate the need for virtual methods. Applying this to other existing OMR components will go a long way toward clarifying our glue requirements and streamlining execution of component functions.

Instead of waiting on template tooling support if there is a way to hide template use from existing tooling that can't handle templates, consider patching as required and pulling this.

ktbriggs commented:
@charliegracie this is ready for review. The most recent commit (96f611c) addresses code review comments from previous commit (af9802a) (including comments relating to object allocation framework). Both of these commits are required. Will squash post-review if requested.

ktbriggs commented:
Template-based solution was abandoned for 2 main reasons:
1. DDR does not provide sufficient support, class template use breaks builds.
2. Class templates cannot have an associated cpp file as the template parameters are not known to the compiler when compiling the cpp file. 
Instead introduced component-delegate pattern, with a GC_ObjectModelDelegate class with interface (methods, method signatures) defined by OMR and language-specific implementation. This provides everything required of the template solution, including inline delegate method implementations, without the limitations.

Testing:
omr.personal build (all platforms): http://vmfarm.ottawa.ibm.com/build_info.php?build_id=328327

x86-64/linux (with J9/compressedrefs, all gc policies, 48 CPUs, 48G heap)
- SPECjvm2008 (derby xml.validation mpegaudio serial compress scimark.lu.large)
- SPECjbb2005
- SPECjbb2015

Total elapsed runtime ~16 hours (~768 CPU-hours). No segfaults, although SPECjbb2015 performed badly due to lack of tuning of SPEC properties, same results would be seen with sdk built from HEAD.

ktbriggs commented:
@charliegracie All review comments above have been address. Issues flagged during desk-side reviews that are out of scope for current issue #296 have been filed as new issues, and changes related to those issues will not be addressed here (see #559, #561). Most recent commit (6264d89) squashes the 3 commits previously reviewed and includes one additional code change (in scope) from most recent review (base GC_ObjectModelBase on MM_BaseVirtual).

ktbriggs commented:
@charliegracie Changes in most recent commit 1e437b4 enable OMR_GC_MODRON_COMPACTION and replace related #error with #message to verify that main commit 69012ad does not itself break OMR builds when OMR_GC_MODRON_COMPACTION is enabled. 

If you are satisfied that successful Travis build for this branch with 1e437b4 proves the point then let me know. I will then reverse the #error -> #message changes and patch the CLI_USING_LAZY_HASHCODE_INSERTION -> OMR_GC_DEFERRED_HASHCODE_INSERTION changes into 69012ad.

Final verification will be all-platforms j9vm build with patched 69012ad and robust tests on VMFarm. I'll also run a series of SPEC benchmarks on the jvmxa6490 SDK from the j9vm build.

ktbriggs commented:
Commit e489acc partially reverses previous commit 4d3382 to restore #error messages and disablement of OMR_GC_COMPACTION. It preserves the macro name change CLI_USING_LAZY_HASHCODE_INSERTION -> OMR_GC_DEFERRED_HASHCODE_INSERTION, which Java must define in its object model. 

This commit should complete this pull request, unless issues are found in SPEC load testing.

ktbriggs commented:
Commit e489acc partially reverses previous commit 4d3382 to restore #error messages and disablement of OMR_GC_COMPACTION. It preserves the macro name change CLI_USING_LAZY_HASHCODE_INSERTION -> OMR_GC_DEFERRED_HASHCODE_INSERTION, which Java must define in its object model. 

This commit should complete this pull request, unless issues are found in SPEC load testing.

charliegracie commented:
Great work Kim! I believe this really simplifies the initial consumption of the GC component.  I am excited about seeing the other changes for ConfigurationLanguageInterface and CollectorLanguageInterface.
